custom_components/tahoma/binary_sensor.py

```diff
@@ -69,55 +99,43 @@ async def async_setup_entry(
     """Set up the TaHoma sensors from a config entry."""
     data = hass.data[DOMAIN][entry.entry_id]
     coordinator = data["coordinator"]
+    entities = []
+
+    key_supported_states = {
+        description.key: description for description in BINARY_SENSOR_DESCRIPTIONS
+    }
+
+    for device in coordinator.data.values():
+        for state in device.states:
+            description = key_supported_states.get(state.name)
+
+            if description:
+                entities.append(
+                    TahomaBinarySensor(
+                        device.deviceurl,
+                        coordinator,
+                        description,
+                    )
+                )
 
-    entities = [
-        TahomaBinarySensor(device.deviceurl, coordinator)
-        for device in data["platforms"][BINARY_SENSOR]
-    ]
     async_add_entities(entities)
 
 
 class TahomaBinarySensor(TahomaEntity, BinarySensorEntity):
     """Representation of a TaHoma Binary Sensor."""
 
+    def __init__(
+        self,
+        device_url: str,
+        coordinator: TahomaDataUpdateCoordinator,
+        description: OverkizBinarySensorDescription,
+    ):
+        """Initialize the device."""
+        super().__init__(device_url, coordinator)
+        self.entity_description = description
+
     @property
     def is_on(self):
         """Return the state of the sensor."""
-
-        return (
-            self.select_state(
-                CORE_ASSEMBLY_STATE,
-                CORE_BUTTON_STATE,
-                CORE_CONTACT_STATE,
-                CORE_GAS_DETECTION_STATE,
-                CORE_OCCUPANCY_STATE,
-                CORE_OPENING_STATE,
-                CORE_OPEN_CLOSED_TILT_STATE,
-                CORE_RAIN_STATE,
-                CORE_SMOKE_STATE,
-                CORE_THREE_WAY_HANDLE_DIRECTION_STATE,
-                CORE_VIBRATION_STATE,
-                CORE_WATER_DETECTION_STATE,
-                IO_VIBRATION_DETECTED_STATE,
-            )
-            in [STATE_OPEN, STATE_PERSON_INSIDE, STATE_DETECTED, STATE_PRESSED]
-        )
-
-    @property
-    def device_class(self):
-        """Return the class of the device."""
-        return TAHOMA_BINARY_SENSOR_DEVICE_CLASSES.get(
-            self.device.widget
-        ) or TAHOMA_BINARY_SENSOR_DEVICE_CLASSES.get(self.device.ui_class)
-
-    @property
-    def icon(self) -> Optional[str]:
-        """Return the icon to use in the frontend, if any."""
-        if self.device_class == DEVICE_CLASS_WATER:
-            if self.is_on:
-                return ICON_WATER
-            return ICON_WATER_OFF
-
-        icons = {DEVICE_CLASS_GAS: ICON_WAVES, DEVICE_CLASS_RAIN: ICON_WEATHER_RAINY}
-
-        return icons.get(self.device_class)
+        state = self.select_state(self.entity_description.key)
```

Comment:
In such case, what do you think calling directly the get state from pyhoma? I would like to see select_state usage be reduced.

Reply:
Yes, makes sense. Will do this the same as the sensor.